src/go/metrics/collector.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2021 The Magma Authors.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/benbjohnson/clock"
+)
+
+type metricsCollector struct {
+	sender MetricSender
+	stop   chan bool
+}
+
+type MetricsCollectionOptions struct {
+	Delay time.Duration
+	Clock clock.Clock
+}
+
+var collectorOnce sync.Once
+var collectorInstance *metricsCollector
+
+func DefaultMetricsCollectionOpts() *MetricsCollectionOptions {
+	return &MetricsCollectionOptions{
+		Delay: time.Minute,
+		Clock: clock.New(),
+	}
+}
+
+func StartMetricsCollection(options *MetricsCollectionOptions) error {
+	collector := getMetricsCollector()
+	if collector.stop != nil {
+		return errors.New("cannot start metrics collection more than once")
+	}
+	if collector.sender == nil {
+		return errors.New("cannot start metrics collection without registered MetricSender")
+	}
+	collector.stop = make(chan bool)
+
+	go func(collector *metricsCollector) {
+		for {
+			collector.collectMetrics()
+			select {
+			case <-options.Clock.After(options.Delay):
+			case <-collector.stop:
+				return
+			}
+		}
+	}(collector)
+
+	return nil
+}
+
+func StopMetricsCollection() {
+	collector := getMetricsCollector()
+	collector.stop <- true
+}
+
+func RegisterSender(sender MetricSender) error {
```

Comment:
do you expect `Sender` to change? if so, move `Sender` out of `NewCollector` and make it a `SetSender` func

but it looks like no, given that you're erroring when someone tries to set the Sender when it already exists

Reply:
I think somebody might eventually want to send metrics to multiple places (and register multiple `Sender`s), so the naming is reflecting that, though it isn't possible to do at the moment.

Reply:
...Though given the current limitation, I'll just remove `RegisterSender` and just have it included as part of `NewCollector`